index.js

```diff
@@ -4,15 +4,103 @@
 
   /*istanbul ignore next*/
   if(module && typeof module.exports !== 'undefined'){
-    module.exports = f();
+    module.exports = f(require('sanctuary-type-classes'), require('sanctuary-type-identifiers'));
   }else{
-    global.concurrify = f();
+    global.concurrify = f(global.sanctuaryTypeClasses, global.sanctuaryTypeIdentifiers);
   }
 
-}(/*istanbul ignore next*/(global || window || this), function(){
+}(/*istanbul ignore next*/(global || window || this), function(Z, type){
 
   'use strict';
 
-  return function(){};
+  var $alt = 'fantasy-land/alt';
+  var $ap = 'fantasy-land/ap';
+  var $map = 'fantasy-land/map';
+  var $of = 'fantasy-land/of';
+  var $zero = 'fantasy-land/zero';
+  var $$type = '@@type';
+
+  var ordinal = ['first', 'second', 'third', 'fourth', 'fifth'];
+  var show = Z.toString;
+
+  function invalidArgument(it, at, expected, actual){
+    throw new TypeError(
+      it
+      + ' expects its '
+      + ordinal[at]
+      + ' argument to '
+      + expected
+      + '\n  Actual: '
+      + show(actual)
+    );
+  }
+
+  function invalidContext(it, actual){
+    throw new TypeError(
+      it + ' was invoked outside the context of a Concurrently. \n  Called on: ' + show(actual)
+    );
+  }
+
+  function getReprType(Repr){
+    return Repr[$$type] || 'concurrify/' + (Repr.name || 'Anonymous');
+  }
+
+  function computeType(type){
+    var parts = type.split('/');
+    return parts[0] + '/Concurrent' + parts[1];
+  }
+
+  //concurrify :: Applicative m
+  //           => (TypeRep m, () -> m a, (m a, m a) -> m a, (m a, m (a -> b)) -> m b)
+  //           -> Concurrently m
+  return function concurrify(Repr, zero, race, ap){
+
+    var INNERTYPE = getReprType(Repr);
+    var OUTERTYPE = computeType(INNERTYPE);
+
+    function Concurrently(sequential){
+      this.sequential = sequential;
+    }
+
+    function isConcurrently(x){
+      return x instanceof Concurrently || type(x) === OUTERTYPE;
+    }
+
+    function construct(x){
+      if(x.constructor !== Repr)
+        invalidArgument(OUTERTYPE, 0, 'be of type "' + INNERTYPE + '"', type(x));
```

Comment:
Instead of checking for Applicative, I took the much cheaper (and likely more useful) option of checking whether the input was constructed by `Repr`. This builds on the assumption that the input to `concurrify` was correct (done by libs) and the input to `construct` may not be. It also immediately restricts input to the set of values that work with the `zero`, `race` and `ap` functions.